### PR TITLE
feat:  Adding support for altering and overriding overlay actions

### LINKED ISSUE #461: Add permissions API and actionBar override
This ticket covers two features:

1. Adding an API to override the action bar and inject new items
2. Adding a new permissions API to control key parts of functionality for:
  * `drag`: whether the component can be dragged
  * `edit`: whether or not fields are editable (read-only if not)
  * `delete`: whether the component can be deleted
  * `duplicate`: whether the component can be duplicated
  * `insert`: whether the component can be inserted
  * ... TBD

## Tasks

- [x] Export Action Bar #563 
- [x] Add actionBar override #action bar #570
- [x] Implement permissions API #572
- [x] Configure default permissions from (2) above #574
- [x] Implement resolvePermissions API  #581 
- [ ] #579 
- [ ] #580
- [ ] #584 
- [ ] #582
- [ ] Documentation

### Injecting new items via actionBar override

```tsx
import { ActionBar, Puck } from "@measured/puck";

const data = {};

export function Editor() {
  return (
    <Puck
      data={data}
      overrides={{
        // Enable override of entire action bar, not just internal actions
        actionBar: ({ children, item, itemSelector }) => {
          return (
            // <ActionBar> is the default UI
            <ActionBar>
              <div
                style={{
                  display: "flex",
                }}
              >
                <ActionBar.Item>My custom item</ActionBar.Item>
                {children}
            </ActionBar>
          );
        },
      }}
    />
  );
}
```

### Adding `permissions` and `resolvePermissions` APIs

```tsx
import { usePuck } from "@measured/puck";

const data = {};

export function Editor() {
  const permissions = {
    duplicate: false,
    edit: true,
    // A custom permission for the user
    party: false,
  };
  
  return (
    <Puck
      // Specify global permissions
      permissions={permissions}

      config={{
        components: {
          MyComponent: {
            // Basic usage
            permissions: { edit: false },

            // Advanced usage
            resolvePermissions: async ({
              props,
              lastPermissions,
              initialPermissions,
            }) => {
              // Do some async task
              const isAdmin = await getAdmin();
              
              return {
                party: isAdmin
               };
            },
          },
        },
      }}
      data={data}
    />
  );
}
```

### PR DESCRIPTION
**Pull request by xaviemirmon**
Closes #461 

Still todo:

- [ ]  Fix field styling and add classes to stylesheet
- [ ] Refactor shared overrides code to keep things DRY

### CODE REVIEW

**Review by chrisvxd — CHANGES_REQUESTED** [review]
This is great! I've left a bunch of comments, but have also been riffing on this and playing with some more ideas taking the essence of what you've done.

Fundamentally, I think we should:

1. Call this **permissions** instead of **overlayActions** to reflect it can be used for other things (like disabling the form)
2. Avoid putting permission info in the Data object and use the UI state instead to avoid bloating the user payload
3. Enable the user to specify global permissions at the `<Puck>` level
4. Enable the user to specify component permissions at the component level
5. Enable the user to specify instance (local) permissions based on the ID
6. Provide some helpers for getting and setting permissions via `usePuck`
7. Enable overriding of the entire overlayAction bar (which I think should be called the `actionBar` override).

I think we can probably split this work into 2 PRs:

1. The core permissions API and usePuck enhancements
2. The overrides API and associated component exports
 
Here's a fully fledged example that describes what I'm thinking:

```tsx
"use client";

import { ActionBar, Puck, usePuck } from "@/core";

import { Lock, Unlock } from "lucide-react";

const data = {};

export function Editor() {
  const permissions = {
    duplicate: false,
    edit: true,
    // A custom permission for the user
    party: true,
  };

  return (
    <Puck
      // Specify global permissions
      initialPermissions={permissions}
      // Track permissions in UI state to enable modifying via dispatcher
      ui={{
        permissions: {
          // Global permissions
          global: {
            delete: true, // Add the default, as not specified by user
            duplicate: false,
            edit: true,
            party: true,
          },
          // Component permissions, as specified by a component config
          // Inherits global
          component: {
            MyComponent: {
              edit: false,
            },
          },
          // Permissions for specific instances of a component
          // Inherits component
          local: {
            "MyComponent-1234": {
              edit: true,
            },
          },
        },
      }}
      config={{
        components: {
          MyComponent: {
            permissions: {
              edit: false,
            },
          },
        },
      }}
      data={data}
      overrides={{
        // Enable override of entire action bar, not just internal actions
        actionBar: ({ children, item }) => {
          const {
            // New API to get and mutate permissions on UI
            getPermission,
            setPermission,
          } = usePuck();

          const { id, type } = item.props;

          // Get resolved permission, with local first, then component, then global.
          const isEditable = getPermission({ permission: "edit", id, type });

          return (
            // <ActionBar> is the default UI
            <ActionBar>
              <div
                style={{
                  display: "flex",
                }}
              >
                {isAdmin && (
                  <button
                    className={"styles_DraggableComponent-action__LbbWP"}
                    onClick={() => {
                      setPermission(
                        {
                          permission: "edit",
                          id, // Set local permission for this ID
                          type, // Set component permission for this type
                          // Or set global permission if neither specified
                        },
                        !isEditable
                      );
                      // setPermission modifies state internally 👇
                      // dispatch({
                      //   type: "setUi",
                      //   ui: {
                      //     permissions: {
                      //       local: {
                      //         [id]: {
                      //           edit: !isEditable,
                      //         },
                      //       },
                      //     },
                      //   },
                      // });
                    }}
                  >
                    {!isEditable ? <Unlock size={16} /> : <Lock size={16} />}
                  </button>
                )}
              </div>
              {children}
            </ActionBar>
          );
        },
      }}
    />
  );
}
```

File: packages/core/components/Puck/components/Fields/index.tsx [hunk 1]
```diff
@@ -154,135 +156,171 @@ export const Fields = () => {
...
-            } else {
-              currentProps = rootProps;
-            }
-
-            const newProps = {
-              ...currentProps,
-              [fieldName]: value,
-            };
+    <div {...{ inert: !isEditable ? "" : undefined }}>
+      {!isEditable && (
```
[the remark this answers is not in the document]
Comment (xaviemirmon) [reply packages/core/components/Puck/components/Fields/index.tsx:168]: That was my initial approach, but it didn't work on all fields e.g. the External and array on the hero.

[the remark this answers is not in the document]
Comment (chrisvxd) [reply packages/core/components/Puck/components/Fields/index.tsx:168]: If that's the case, I think we should address those as bugs instead

**Comment by chrisvxd** [comment]
Another proposal that differs in two key ways:

1. UI state no longer stores permissions - instead, we let the user handle these state changes
2. The `resolvePermissions` API is added to enable dynamic permissions on a per-component level. Permissions can consequently be applied dynamically based on any prop, not just ID.

```tsx
"use client";

import { ActionBar, Puck, usePuck } from "@/core";

import { Lock, Unlock } from "lucide-react";
import { useState } from "react";

const data = {};

export function Editor() {
  const [permissions] = useState({
    duplicate: false,
    edit: true,
    // A custom permission for the user
    party: true,
  });

  const [myComponentPermissions] = useState({
    edit: true,
    party: false,
  });

  const [localPermissions, setLocalPermissions] = useState<
    Record<string, object>
  >({});

  return (
    <Puck
      // Specify global permissions
      permissions={permissions}
      // Track permissions in UI state to enable modifying via dispatcher
      config={{
        components: {
          MyComponent: {
            // Basic usage
            permissions: myComponentPermissions,

            // Advanced usage
            resolvePermissions: ({
              props,
              lastPermissions,
              initialPermissions,
            }) => {
              if (localPermissions[props.id]?.edit === false) {
                return {
                  ...myComponentPermissions,
                  edit: false,
                };
              }

              return myComponentPermissions;
            },
          },
        },
      }}
      data={data}
      overrides={{
        // Enable override of entire action bar, not just internal actions
        actionBar: ({ children, item }) => {
          // New API to get resolved permissions
          const { getPermission } = usePuck();

          const { id, type } = item.props;

          // Get resolved permission, with local first, then component, then global.
          const isEditable = getPermission({ permission: "edit", type });

          return (
            // <ActionBar> is the default UI
            <ActionBar>
              <div
                style={{
                  display: "flex",
                }}
              >
                {isAdmin && (
                  <ActionBar.Action
                    onClick={() => {
                      setLocalPermissions((obj) => ({
                        ...obj,
                        [id]: { ...obj[id], edit: !isEditable },
                      }));
                    }}
                  >
                    {!isEditable ? <Unlock size={16} /> : <Lock size={16} />}
                  </ActionBar.Action>
                )}
              </div>
              {children}
            </ActionBar>
          );
        },
      }}
    />
  );
}
```

**Pull request closed without merging**